src/librustc/ty/layout.rs

```diff
@@ -1086,10 +1148,17 @@ impl<'a, 'gcx, 'tcx> Layout {
                             });
                         }
 
+                        let st = Struct::new(dl,
+                            variants[discr].iter().map(|ty| ty.layout(infcx)),
+                            hint, false, ty)?;
+
+                        // We have to fix the last element of path here as only we know the right value.
+                        let mut i = *path.last().unwrap();
+                        i = st.field_index[i as usize];
+                        *path.last_mut().unwrap() = i;
```

Comment:
That wouldn't be an improvement because the borrow lasts too long.  It would have to be wrapped in braces and indented further or I'd have to use mem::forget.  We can't push the 0 onto the Vec while `i` exists.